presto-main/src/main/java/com/facebook/presto/operator/scalar/MapTransformKeyFunction.java

```diff
@@ -95,39 +119,142 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
+            loadKeyElement = new BytecodeBlock()
+                    .append(keyElement.set(constantNull(keyJavaType)))
+                    .append(throwNullKeyException);
         }
-        return resultBuilder.build();
+
+        SqlTypeBytecodeExpression valueSqlType = constantType(binder, valueType);
+        BytecodeNode loadValueElement;
+        if (!valueType.equals(UNKNOWN)) {
+            loadValueElement = new IfStatement()
+                    .condition(block.invoke("isNull", boolean.class, add(position, constantInt(1))))
+                    .ifTrue(valueElement.set(constantNull(valueJavaType)))
+                    .ifFalse(valueElement.set(valueSqlType.getValue(block, add(position, constantInt(1))).cast(valueJavaType)));
+        }
+        else {
+            // make sure invokeExact will not take uninitialized keys during compile time
+            loadValueElement = new BytecodeBlock().append(valueElement.set(constantNull(valueJavaType)));
+        }
+
+        SqlTypeBytecodeExpression transformedKeySqlType = constantType(binder, transformedKeyType);
+        BytecodeNode writeKeyElement;
+        BytecodeNode throwDuplicatedKeyException;
+        if (!transformedKeyType.equals(UNKNOWN)) {
+            writeKeyElement = new BytecodeBlock()
+                    .append(transformedKeyElement.set(function.invoke("invokeExact", transformedKeyJavaType, keyElement, valueElement)))
+                    .append(new IfStatement()
+                            .condition(equal(transformedKeyElement, constantNull(transformedKeyJavaType)))
+                            .ifTrue(throwNullKeyException)
+                            .ifFalse(new BytecodeBlock()
+                                    .append(constantType(binder, transformedKeyType).writeValue(blockBuilder, transformedKeyElement.cast(transformedKeyType.getJavaType())))
+                                    .append(valueSqlType.invoke("appendTo", void.class, block, add(position, constantInt(1)), blockBuilder))));
+
+            // make sure getObjectValue takes a known key type
+            throwDuplicatedKeyException = new BytecodeBlock()
+                    .append(newInstance(
+                            PrestoException.class,
+                            getStatic(INVALID_FUNCTION_ARGUMENT.getDeclaringClass(), "INVALID_FUNCTION_ARGUMENT").cast(ErrorCodeSupplier.class),
+                            invokeStatic(
+                                    String.class,
+                                    "format",
+                                    String.class,
+                                    constantString("Duplicate keys (%s) are not allowed"),
+                                    newArray(type(Object[].class), ImmutableList.of(transformedKeySqlType.invoke("getObjectValue", Object.class, session, blockBuilder.cast(Block.class), position))))))
+                    .throwObject();
+        }
+        else {
+            // make sure the blocks are initialized
```

Comment:
java requires a bytecode node to be initialized under any case; if transformedKeyType is unknown, essentially there is nothing right to be initialized for writeKeyElement because that's an error. So we just make it an exception.